heksher/api/v1/rules.py

```diff
@@ -0,0 +1,186 @@
+                                 status_code=status.HTTP_400_BAD_REQUEST)
+
+    existing_rule = await app.db_logic.get_rule_id(input.setting, input.feature_values)
+    if existing_rule:
+        return PlainTextResponse('rule already exists', status_code=status.HTTP_409_CONFLICT)
+
+    new_id = await app.db_logic.add_rule(input.setting, input.value, input.metadata, input.feature_values)
+    await app.db_logic.touch_setting(input.setting)
+
+    return AddRuleOutput(rule_id=new_id)
+
+
+class QueryRulesInput(ORJSONModel):
+    setting_names: List[SettingName] = Field(description="a list of setting names to return the rules for")
+    context_features_options: Dict[ContextFeatureName, List[ContextFeatureValue]] = Field(
+        description="a mapping of context features and possible values. Any rule with an exact-match condition not in"
+                    " this mapping will not be returned"
+    )
+    cache_time: Optional[datetime] = Field(None, description="if provided, any settings that have not been changed"
+                                                             " since this time will be ignored")
+    include_metadata: bool = Field(False, description="whether to load and include the metadata of each rule in"
+                                                      " the results")
+
+
+class Rule(ORJSONModel):
+    setting: str = Field(description="the setting the rule applies to")
+    value: Any = Field(description="the value of the setting in contexts where the rule matches")
+    context_features: List[Tuple[str, str]] = Field(description="a list of exact-match conditions for the rule")
+
+
+class RuleWithMetadata(Rule):
+    metadata: Dict[str, Any] = Field(description="the metadata of the rule, if requested")
+
+
+class QueryRulesOutput(ORJSONModel):
+    rules: List[Rule] = Field(description="the rules that matched the query")
+    included_settings: List[str] = Field(description="the settings that were not filtered out of the query")
+
+
+class QueryRulesOutputWithMetadata(ORJSONModel):
+    rules: List[RuleWithMetadata] = Field(description="the rules that matched the query")
+    included_settings: List[str] = Field(description="the settings that were not filtered out of the query")
+
+
+@router.get('/query', response_model=Union[QueryRulesOutputWithMetadata, QueryRulesOutput])
+async def query_rules(input: QueryRulesInput, app: HeksherApp = application):
+    """
+    query settings for rules for a specific set of potential contexts
+    """
+    not_context_features = await app.db_logic.get_not_context_features(input.context_features_options)
+    if not_context_features:
+        return PlainTextResponse(f'the following are not valid context features: {not_context_features}',
+                                 status_code=status.HTTP_404_NOT_FOUND)
+
+    not_settings = await app.db_logic.get_not_settings(input.setting_names)
```

Comment:
this whole not_context_features and not_settings is very odd. do you mean invalid_settings?

Reply:
That name would be misleading, as validation occurs on the routing level, not the DB level. not_settings and not_context_features should check for whether a setting or CF **currently** exists in the DB

Reply:
I'm pretty sure you can find a better name